core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNode.java

```diff
@@ -158,55 +171,73 @@ public DiscoveryNode(String nodeName, String nodeId, TransportAddress address, M
+        this(nodeName, nodeId, hostName, hostAddress, address, copyAttributes(attributes), roles, version);
     }
 
     /**
      * Creates a new {@link DiscoveryNode}.
      * <p>
-     * <b>Note:</b> if the version of the node is unknown {@link Version#minimumCompatibilityVersion()} should be used for the current version.
-     * it corresponds to the minimum version this elasticsearch version can communicate with. If a higher version is used
+     * <b>Note:</b> if the version of the node is unknown {@link Version#minimumCompatibilityVersion()} should be used for the current
+     * version. it corresponds to the minimum version this elasticsearch version can communicate with. If a higher version is used
      * the node might not be able to communicate with the remove node. After initial handshakes node versions will be discovered
      * and updated.
      * </p>
      *
      * @param nodeName    the nodes name
-     * @param nodeId      the nodes unique id.
-     * @param hostName    the nodes hostname
-     * @param hostAddress the nodes host address
-     * @param address     the nodes transport address
-     * @param attributes  node attributes
-     * @param version     the version of the node.
+     * @param nodeId      the nodes unique id
+     * @param node        the original node to copy all the information from
      */
-    public DiscoveryNode(String nodeName, String nodeId, String hostName, String hostAddress, TransportAddress address, ImmutableOpenMap<String, String> attributes, Version version) {
+    public DiscoveryNode(String nodeName, String nodeId, DiscoveryNode node) {
+        this(nodeName, nodeId, node.hostName, node.hostAddress, node.address, copyAttributes(node.attributes), node.roles, node.version);
+    }
+
+    private DiscoveryNode(String nodeName, String nodeId, String hostName, String hostAddress, TransportAddress address,
+                          ImmutableOpenMap.Builder<String, String> attributesBuilder, Set<Role> roles, Version version) {
         if (nodeName != null) {
             this.nodeName = nodeName.intern();
+        } else {
+            this.nodeName = "";
         }
-        ImmutableOpenMap.Builder<String, String> builder = ImmutableOpenMap.builder();
-        for (ObjectObjectCursor<String, String> entry : attributes) {
-            builder.put(entry.key.intern(), entry.value.intern());
-        }
-        this.attributes = builder.build();
         this.nodeId = nodeId.intern();
         this.hostName = hostName.intern();
         this.hostAddress = hostAddress.intern();
         this.address = address;
-        this.version = version;
+        if (version == null) {
+            this.version = Version.CURRENT;
+        } else {
+            this.version = version;
+        }
+        this.attributes = attributesBuilder.build();
+        //verify that no node roles are being provided as attributes
+        boolean assertEnabled = false;
+        assert assertEnabled = true;
```

Comment:
I added the message....but the function magic didn't work for me...do we have that anywhere in our codebase already?